Rolling update failure thresholds and rollback

aaronlehmann opened the pull request:
This adds a previous spec to the Serivce object to support rollback.
It also adds new failure actions and update states corresponding to
rollback situations. Finally, it adds a threshold determining whether to
invoke the failure action.

I also thought about adding a timeout on task startup, so that if a
task does not reach RUNNING within the given amount of time, it
counts as a failure and is taken account when evaluating if the
update is above or below its failure threshold. However, a timeout
on task startup shouldn't be specific to rolling updates, and is
somewhat difficult to handle correctly across all the possible
situations (new tasks, updated tasks, restarted tasks, leader
failover...). I decided to leave this for the future.

cc @dongluochen @stevvooe @aluzzardi

dongluochen commented on api/types.proto:
```diff
@@ -281,15 +281,32 @@ message UpdateConfig {
+	//
+	// If the failure action is CONTINUE, there is no effect.
+	// If the failure action is PAUSE, no more tasks will be updated until
+	// another update is started.
```
Does the next update start with failing task?

aaronlehmann replied:
I'm not sure I understand the question. What I mean here is that no more tasks will be updated until the user changes the `ServiceSpec`.

dongluochen replied:
When user changes the `ServiceSpec` again, i.e., another update is started, does updater start with the failing fraction first?

aaronlehmann replied:
If the `ServiceSpec` changes, it's considered a fresh update. Only tasks with the new spec will count towards the number of failures.

dongluochen replied:
I think we should keep service healthy as much as possible. If previous rollout PAUSED because of failure, the next rollout should resolve those failures first.

aaronlehmann replied:
I think we could do this by sorting the instances by health and doing the updates in that order. It shouldn't have an impact on the protobufs, though.

dongluochen replied:
Agree.

stevvooe commented on api/objects.proto:
```diff
@@ -57,6 +57,9 @@ message Service {
 	ServiceSpec spec = 3 [(gogoproto.nullable) = false];
 
+	// LastSpec is the previous service spec that was in place before "Spec".
+	ServiceSpec last_spec = 6;
```
Nit: "previous_spec".

stevvooe commented on api/types.proto:
```diff
@@ -299,20 +316,27 @@ message UpdateStatus {
 		COMPLETED = 3;
-		// TODO(aaronl): add ROLLING_BACK, ROLLED_BACK as part of
-		// rollback support.
+		ROLLING_BACK = 4;
```
`ROLLBACK_STARTED`

dongluochen commented on api/types.proto:
```diff
@@ -299,20 +316,27 @@ message UpdateStatus {
 
+	// StartedRollingBackAt is the time at which the update started rolling
+	// back to the previous version.
+	Timestamp started_rolling_back_at = 5;
```
`rollback_started_at` - I prefer keeping rollback as a word.

dongluochen replied:
Why skip 4.

aaronlehmann replied:
4 is below. I wanted to group the timestamps together.

stevvooe commented on api/types.proto:
```diff
@@ -299,20 +316,27 @@ message UpdateStatus {
+	// StartedRollingBackAt is the time at which the update started rolling
+	// back to the previous version.
+	Timestamp started_rolling_back_at = 5;
+
```
Can we include a progress message in here to support `running/replicas` in command output? That will let us get rid of some extra queries to task sets and make that  more accurate.

aaronlehmann replied:
The difficulty with this is that the service object would get updated frequently. Right now this contributes to some issues like #1379. If we fix that stuff, I think including a progress message would be reasonable.

stevvooe replied:
@aaronlehmann #1379 is a feature!

This is probably more than just a message. We'd want to know the number of satisfied slots vs. the number of satisfiable slots.

aaronlehmann replied:
From a user's perspective, I don't think it's fair to reject a service update because the number of satisfied slots changed while the "docker service update" command was running - especially because the user has no non-racy way of doing a check-and-set based on this information. I think we should either put in automatic retries or a spec version to avoid this user-facing issue.

stevvooe commented on api/types.proto:
```diff
@@ -281,15 +281,32 @@ message UpdateConfig {
+	// an update before the failure action is invoked. Any task created by
+	// the current update which ends up in one of the states REJECTED,
+	// COMPLETED or FAILED while the update is still running is counted as
+	// a failure. The number of failures is divided by the number of tasks
```
How do you know which failures belong to which update?

aaronlehmann replied:
> How do you know which failures belong to which update?

My plan is to do it by comparing task specs. If a task failed, and its spec matched the one we're updating to, then we would conclude the task was started by the current update.

I think this is the right approach because it's very simple and doesn't require keeping any local state that would be lost during a manager failover.

stevvooe replied:
Then failures of a previous Spec may apply to this ratio unintentionally. This can happen if one tries an update, has a temporary condition, then retries the update with the same Spec.

aaronlehmann replied:
I considered that case, but I think I have a way to work around it. If `Spec` and `PreviousSpec` match, we can disregard failures entirely, because it wouldn't make sense to pause or roll back in this case.

stevvooe replied:
What if they aren't? There are conditions where you can never change to a certain Spec because it encountered too many failures.

aaronlehmann replied:
That's a good point. What if we were to version specs? For example, we could keep a counter inside the Service object of how many times it has been updated, and copy that counter to the Task objects. By comparing this version number, we could distinguish tasks created by different updates, even if the specs match.

stevvooe replied:
I'm not sure versioning the specs is the right solution. Really, we are interested in whether or not the task failures belong to the current ongoing update. Basically, do they belong to the same "generator". So, that would be versioning the update.

Albeit, this is slightly equivalent to versioning the specifications, but does so without the implications of doing so.

aaronlehmann replied:
I think versioning the update is what I meant to say.

stevvooe commented:
> I also thought about adding a timeout on task startup, so that if a
task does not reach RUNNING within the given amount of time, it
counts as a failure and is taken account when evaluating if the
update is above or below its failure threshold.

This should be taken care of by the executor. If the task doesn't start in a certain amount of time we report failure.

dongluochen commented:
LGTM

aaronlehmann commented:
Updated to record the service version (from `Meta`) in the `UpdateStatus`, and propagate this version information to tasks that are created by the update. This will allow us to match up tasks with the update that created them.

stevvooe commented on manager/orchestrator/replicated.go:
```diff
@@ -132,6 +132,7 @@ func newTask(cluster *api.Cluster, service *api.Service, instance uint64) *api.T
 		ServiceAnnotations: service.Spec.Annotations,
 		Spec:               service.Spec.Task,
 		ServiceID:          service.ID,
+		ServiceVersion:     service.Meta.Version,
```
Does this need to be set from `UpdateStatus`?

aaronlehmann replied:
Yeah, that was the intent but I messed this up. Fixing now.

aaronlehmann replied:
> Does this need to be set from `UpdateStatus`?

Fixed.

aaronlehmann commented:
Rebased. ping @aluzzardi

aaronlehmann commented:
Discussed rollbacks with @aluzzardi and came up with a few conclusions:

- Counting any failures that happen during the update period is too non-deterministic. We're thinking of supporting a configurable monitoring period that applies on a per-task basis. In other words, if this is set to 30 seconds, a task failing during the first 30 seconds after it's updated will count as a failure, and after that it will not count as a failure. This is similar to only watching for failures during the "update delay" that optionally happens after a task is updated, but it lets us decouple the pause between updates with the monitoring period, and put in place a more useful default for the monitoring period (as opposed to the 0s default for pausing between updates).

- If a task fails during the update, and then continues to fail as it's restarted, that should only count as a single failure.

- Looking at failed tasks in the data store isn't a good way of counting failures. This is too sensitive to task reaper settings, among other things. We should either count failures with local state, a counter in `UpdateStatus` on the service object, or some other approach that doesn't rely on historic task objects.

- In the model where we have a configurable monitoring period that applies to each task separately, it's not clear how to make this work seamlessly when a manager fails over. Keeping a failure count in the service object is one approach, but even this is not bulletproof (a manager can fail while attempting to update this counter). Maintaining this extra state also increases complexity. For now, it may be best not to make any guarantees about accurate failure accounting across leaders, and consider adding this as a future enhancement. The worst case scenario is that a leader failover resets the failure count during an update, and inhibits a rollback that otherwise would have happened. This is a corner case that I don't know if many people care about (however, as a counterpoint, I don't really like adding special cases where things behave differently depending whether the leader changes).

aaronlehmann commented:
I've updated this according to the comments above. I removed the `ServiceVersion` stuff, because I don't think it will be necessary under the model where we monitor tasks individually for specific time  periods, instead of watching for failures of any tasks associated with the current version of the service. I also added `MonitoringPeriod` and clarified some comments to more closely match the mode of operation discussed above.

@aluzzardi PTAL

aaronlehmann commented:
Added a commit to implement `AllowedFailureFraction` and `MonitoringPeriod` in the updater. Currently this uses only local state so the failure count will reset if there's a leader failover. However, I made it evaluate the failure fraction as `totalFailures / len(dirtySlots)`, so if another leader takes over partway through the update, the correct proportion of task failures will trigger the failure action. I'm wondering whether this is a good idea or not, since it means that if the leader changes before updating the last task of many, and that last task fails to update, it could trigger a rollback.

Currently this change just deals with pausing the update because rollbacks aren't implemented yet. Implementing rollbacks as a next step is relatively straightforward.

aaronlehmann commented:
Added a commit that implements rollback and adds a unit test for it. PTAL.

aluzzardi commented on manager/orchestrator/coverage.out:
```diff
@@ -0,0 +1,468 @@
+mode: set
```
This file was probably added by mistake

aluzzardi commented:
I was wondering about this use case:
- Start a rolling update. `spec=2 previous_spec=1`
- In the middle of it, do another update. `spec=3 previous_spec=2`
- `3` is broken - we'll rollback to `2`, no questions asked
- turns out 2 was broken too - we just didn't notice it because we didn't finish the rollout

To mitigate this problem, what if we update `previous_spec` ONLY if an update is not already in progress? In the previous example, we would have kept `spec=3 previous_spec=1`

aaronlehmann commented:
That seems reasonable. I don't know which behavior would be least surprising though.

aluzzardi commented on api/types.proto:
```diff
@@ -281,15 +281,40 @@ message UpdateConfig {
+	// from its creation, this counts as a failure. If it fails after
+	// MonitoringPeriod, it does not count as a failure. If
+	// MonitoringPeriod is unspecified, a default value will be used.
+	Duration monitoring_period = 4;
```
nit: We use `Duration delay` for the other duration and those two work hand in hand, yet they have different a naming "scheme".

What about `monitor` or `watch`?

I'm thinking of consistency also for flag names, e.g. `service update --update-delay` vs  `--update-watch` `--update-monitor`, `--update-monitoring-period` (this PR), ...

aaronlehmann replied:
Changed to `monitor`.

Feel like this might be a little terse, but I see the point about flag names.

aluzzardi commented on api/types.proto:
```diff
@@ -281,15 +281,40 @@ message UpdateConfig {
+	// rollbacks? Should there be a failure action for rollbacks (to allow
+	// them to do something other than pause when the rollback encounters
+	// errors)?
+	float allowed_failure_fraction = 5;
```
I'm going back and forth with this - I don't have a favorite side but it's still worth talking about: Fraction or absolute numbers? (e.g. `tolerated_failures`).

- Cons: They don't scale linearly with the number of instances. We'd probably have to use `0` as a default.
- Pros: More deterministic and simple. e.g. with a fraction it's hard to say you only tolerate one failure, etc.

Thoughts?

TBH, pausing on first failure seems like a good default.

aaronlehmann replied:
I kind of like fractions but I can't see a particularly strong argument for one over the other.

Pausing on the first failure is very easy to do with either a fraction or an absolute threshold (`tolerated_failures` = 0 or `allowed_failure_fraction` = 0.0).

aaronlehmann replied:
Actually, here's an argument for (or against) using a fraction. Right now we're not tallying failed tasks in raft, so the failures are tracked as local state. If you have a leader failover midway through the update, and then all of a sudden something goes wrong and your new tasks start failing, the new leader won't pause or roll back the update until it counts enough failures to hit the threshold. But if the threshold is based on a fraction, the number of failures it needs to observe will be proportionally less because it's handling fewer tasks. Otherwise there could be situations where pausing or rolling back is impossible because not enough tasks are being updated at once.

aluzzardi commented on api/types.proto:
```diff
@@ -299,20 +324,27 @@ message UpdateStatus {
 
+	// RollbackStartedAt is the time at which the update started rolling
+	// back to the previous version.
+	Timestamp rollback_started_at = 5;
```
Do we need `paused_at` for consistency?

aluzzardi replied:
Also, we had this conversation a while ago (at the same of adding pause), don't remember the outcome but: does it make sense to add timestamp fields every time we add a new state?

Wouldn't it help to have a list of `UpdateStatus`es and each time we transition to a new state, we add another item in order to keep history?

The same would apply to task status, node status, and so on.

For instance, if we had this, UX wise instead of displaying events like this:
```
Started At: 2016-09-01T00:19:32
Rollback At: 2016-09-01T00:19:34
Completed At: 2016-09-01T00:19:36
```

We could do something like this:
```
Events:
TIME                       EVENT               ERROR
2016-09-01T00:19:32        Update Started      
2016-09-01T00:19:34        Rollback Started    Unable to find image "foo"
2016-09-01T00:19:36        Rollback Completed      

```

aaronlehmann replied:
Yeah, kind of wish we had done this.

I think we can't remove the existing fields. We could add a new `repeated` field to supersede them, but we'd still have to populate the existing timestamp fields when applicable.

So we could do this, but we'd end up with a weird set of fields in the protobuf because we'd have `started_at` and `completed_at` for backward compatibility, and also a repeated field that lists generic events. Do you think it's worth it?

aaronlehmann replied:
@aluzzardi: I removed `RollbackStartedAt`. PTAL.

aluzzardi commented on api/types.proto:
```diff
@@ -299,20 +324,27 @@ message UpdateStatus {
-		// TODO(aaronl): add ROLLING_BACK, ROLLED_BACK as part of
-		// rollback support.
+		ROLLBACK_STARTED = 4;
+		ROLLBACK_PAUSED = 5; // if a rollback fails
```
Is a rollback ever going to fail?

aaronlehmann replied:
A rollback can fail if it hits the failure threshold when trying to create tasks with the previous spec. If this happens, it goes into the `ROLLBACK_PAUSED` state.

Note there's a TODO above related to this:

```
// TODO(aaronl): Should there be a separate failure threshold for
// rollbacks? Should there be a failure action for rollbacks (to allow
// them to do something other than pause when the rollback encounters
// errors)?
```

aluzzardi commented:
Overall LGTM.

I have a few points I'd like to make regarding some preferences. I don't have strong arguments for them, except perhaps n2: I do really think we shouldn't add more timestamps.

1) Threshold: Should this be an absolute number? I don't have a strong opinion, absolute numbers seem more "predictable" to me (from a user's POI) but I can see the appeal for a fractional threshold

2) Status: I'd avoid adding new timestamps into `UpdateStatus`. Instead, ideally, a Service should have a history (service created, service scaled, service updated, ...). This allows the user to see what happened when. `UpdateStatus` is only there to provide real time update information (see https://github.com/docker/swarmkit/pull/1380#discussion_r77097133). I'd go as far as saying that there could be a single timestamp for UpdateStatus.

3) Auto rollback vs manual rollback. The bulk of the PR design LGTM (thresholds, previous spec, ...). The one thing I'm wondering about is: Should we do automatic rollback at all or should we just provide a CLI based `service update --rollback`? These changes allow the implementation client side (by having the client inspect `UpdateStatus` and push `previous_spec` back). A few reasons for this:

3.a) It's "more" user predictable: A rollback will look to the user as if nothing happened. Either he'll think the update didn't run or maybe she won't notice that we rolled back (tough to notice it's still running a hash rather than another one, and at some point the update was in progress)

3.b) It's more configurable. Going 'back' manually provides the same options as going forward: failure action, parallelism, delay ... Maybe if a deployment goes bad you want to go back in lockstep (parallelism = 0) with no delay.

3.c) If we have manual rollback, we can always add auto rollback later. We can't take it away though

aaronlehmann commented:
> 3) Auto rollback vs manual rollback. The bulk of the PR design LGTM (thresholds, previous spec, ...). The one thing I'm wondering about is: Should we do automatic rollback at all or should we just provide a CLI based service update --rollback? These changes allow the implementation client side (by having the client inspect UpdateStatus and push previous_spec back). A few reasons for this:

> 3.a) It's "more" user predictable: A rollback will look to the user as if nothing happened. Either he'll think the update didn't run or maybe she won't notice that we rolled back (tough to notice it's still running a hash rather than another one, and at some point the update was in progress)

> 3.b) It's more configurable. Going 'back' manually provides the same options as going forward: failure action, parallelism, delay ... Maybe if a deployment goes bad you want to go back in lockstep (parallelism = 0) with no delay.

> 3.c) If we have manual rollback, we can always add auto rollback later. We can't take it away though

I'm almost convinced but I have a few questions/comments about manual rollback.

1) Implementing rollback on the client side would mean that "previous spec" becomes the spec from the aborted update, so running `--rollback` multiple times would alternate between the two specs. Is that the behavior we want?

2) After a successful rollback, the update status would show "update complete". This seems slightly confusing because from the user's perspective the update wasn't completed - it was rolled back.

If we decide to go with client-side rollback, I'm wondering about the possibility of leaving the server-side rollback parts of this PR in place so we can continue to experiment with it, but not exposing the feature through engine-api, so we have the option in the future to remove it. What do you think?

aaronlehmann commented:
Updated PR to mark automatic rollback as experimental. The integration won't expose it for now. It will provide a way to start a rollback manually, instead (`--rollback`).
